File: packages/cli/src/task-display.ts
/** Shared pure task-display logic (Phase 2e-i) — the Claude-Code-tree rendering rules, written
 *  ONCE here and mirrored byte-for-byte in `apple/Norma/Sources/ChatContent/TaskDisplay.swift`.
 *  Task 3 (window) and Task 4 (CLI) each consume this; nothing renders here. Every function is
 *  pure — no I/O, no ANSI, no SwiftUI — so both sides can be unit-tested against the SAME
 *  fixtures (see `test/task-display.test.ts` / `TaskDisplayTests.swift`) and stay in lockstep. */

export interface TaskRow {
  id: string;
  subject: string;
  status: string;
  activeForm?: string;
  startedTs?: number;
}

export interface TaskDisplayResult {
  rows: TaskRow[];
  collapsedCompletedCount: number;
}

/** Sort rank per status: in_progress first, pending (and any unrecognized status) next,
 *  completed last. A plain `Array.prototype.sort` by rank alone is NOT guaranteed stable across
 *  engines by spec pre-ES2019, but more importantly the Swift twin can't rely on it either — so
 *  both sides sort by `(rank, originalIndex)`, which is stable by construction. */
function rankFor(status: string): number {
  if (status === "in_progress") return 0;
  if (status === "completed") return 2;
  return 1;
}

export function sortTasksForDisplay(tasks: TaskRow[]): TaskRow[] {
  return tasks
    .map((task, index) => ({ task, index }))
    .sort((a, b) => {
      const rankDiff = rankFor(a.task.status) - rankFor(b.task.status);
      if (rankDiff !== 0) return rankDiff;
      return a.index - b.index;
    })
    .map(({ task }) => task);
}

/** `sorted` is assumed already in `sortTasksForDisplay` order. Keeps every non-completed row plus
 *  the first `showCompleted` completed rows (in that sorted order); the rest are collapsed into a
 *  count. */
export function collapseCompleted(sorted: TaskRow[], showCompleted: number = 3): TaskDisplayResult {
  const rows: TaskRow[] = [];
  let totalCompleted = 0;
  let keptCompleted = 0;
  for (const task of sorted) {
    if (task.status !== "completed") {
      rows.push(task);
      continue;
    }
    totalCompleted += 1;
    if (keptCompleted < showCompleted) {
      rows.push(task);
      keptCompleted += 1;
    }
  }
  return { rows, collapsedCompletedCount: Math.max(0, totalCompleted - showCompleted) };
}

/** "7 tasks (5 done, 1 in progress, 1 open)" — the CC count header. "open" = pending + any
 *  unrecognized status. Lockstep with TaskDisplay.swift's taskCountsLine. */
export function taskCountsLine(tasks: { status: string }[]): string {
  const done = tasks.filter((t) => t.status === "completed").length;
  const inProgress = tasks.filter((t) => t.status === "in_progress").length;
  const open = tasks.length - done - inProgress;
  return `${tasks.length} tasks (${done} done, ${inProgress} in progress, ${open} open)`;
}

export function taskGlyph(status: string): string {
  if (status === "in_progress") return "■";
  if (status === "completed") return "✓";
  return "☐";
}

/** ms → "14s" / "2m 3s" / "1h 4m". Matches `TaskDisplay.swift`'s `formatElapsed` exactly on the
 *  shared fixtures (14000→"14s", 123000→"2m 3s", 3840000→"1h 4m"). */
export function formatElapsed(ms: number): string {
  const totalSeconds = Math.floor(ms / 1000);
  if (totalSeconds < 60) return `${totalSeconds}s`;
  if (totalSeconds < 3600) {
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = totalSeconds % 60;
    return `${minutes}m ${seconds}s`;
  }
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  return `${hours}h ${minutes}m`;
}

/** n<1000 → integer string ("842"); else one decimal + k/M ("10.6k", "1.2M"). Matches
 *  `TaskDisplay.swift`'s `String(format: "%.1f", Double(n)/1000)` — both always keep exactly one
 *  decimal digit above the 1000 threshold (never strip a trailing ".0"). */
export function formatTokens(n: number): string {
  // Task-1 review fix: one-decimal via INTEGER round-half-up, NOT toFixed/%.1f — those diverge at
  // exact binary ties (n mod 1000 == 250 → toFixed gives 1.3k, Swift %.1f round-half-to-even gives
  // 1.2k), breaking TS↔Swift lockstep on ordinary token counts. `Math.floor((x + half) / step)` is
  // identical to Swift's integer `(x + half) / step`.
  const oneDecimal = (tenths: number) => `${Math.floor(tenths / 10)}.${tenths % 10}`;
  if (n < 1000) return String(n);
  if (n < 1_000_000) return `${oneDecimal(Math.floor((n + 50) / 100))}k`;
  return `${oneDecimal(Math.floor((n + 50_000) / 100_000))}M`;
}

File: apple/Norma/Sources/ChatContent/TaskDisplay.swift
import Foundation

/// Shared pure task-display logic (Phase 2e-i) — the Claude-Code-tree rendering rules, written
/// ONCE here and mirrored byte-for-byte in `packages/cli/src/task-display.ts`. Task 3 (window)
/// and Task 4 (CLI) each consume this; nothing renders here. Every function is pure — no I/O, no
/// SwiftUI — so both sides can be unit-tested against the SAME fixtures (see
/// `TaskDisplayTests.swift` / `test/task-display.test.ts`) and stay in lockstep.

struct TaskRow: Equatable {
    let id: String
    let subject: String
    let status: String
    let activeForm: String?
    let startedTs: Int?
}

struct TaskDisplayResult: Equatable {
    let rows: [TaskRow]
    let collapsedCompletedCount: Int
}

/// Sort rank per status: in_progress first, pending (and any unrecognized status) next,
/// completed last.
private func rank(for status: String) -> Int {
    if status == "in_progress" { return 0 }
    if status == "completed" { return 2 }
    return 1
}

/// Swift's `sort`/`sorted` is NOT guaranteed stable, so ties (equal rank) sort by original index —
/// stable by construction, matching the TS twin's `(rank, originalIndex)` comparator.
func sortTasksForDisplay(_ tasks: [TaskRow]) -> [TaskRow] {
    tasks.enumerated()
        .sorted { a, b in
            let rankDiff = rank(for: a.element.status) - rank(for: b.element.status)
            if rankDiff != 0 { return rankDiff < 0 }
            return a.offset < b.offset
        }
        .map(\.element)
}

/// `sorted` is assumed already in `sortTasksForDisplay` order. Keeps every non-completed row plus
/// the first `showCompleted` completed rows (in that sorted order); the rest are collapsed into a
/// count.
func collapseCompleted(_ sorted: [TaskRow], showCompleted: Int = 2) -> TaskDisplayResult {
    var rows: [TaskRow] = []
    var totalCompleted = 0
    var keptCompleted = 0
    for task in sorted {
        if task.status != "completed" {
            rows.append(task)
            continue
        }
        totalCompleted += 1
        if keptCompleted < showCompleted {
            rows.append(task)
            keptCompleted += 1
        }
    }
    return TaskDisplayResult(rows: rows, collapsedCompletedCount: max(0, totalCompleted - showCompleted))
}

func taskGlyph(_ status: String) -> String {
    if status == "in_progress" { return "■" }
    if status == "completed" { return "✓" }
    return "☐"
}

/// ms → "14s" / "2m 3s" / "1h 4m". Matches `task-display.ts`'s `formatElapsed` exactly on the
/// shared fixtures (14000→"14s", 123000→"2m 3s", 3840000→"1h 4m").
func formatElapsed(_ ms: Int) -> String {
    let totalSeconds = ms / 1000
    if totalSeconds < 60 { return "\(totalSeconds)s" }
    if totalSeconds < 3600 {
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return "\(minutes)m \(seconds)s"
    }
    let hours = totalSeconds / 3600
    let minutes = (totalSeconds % 3600) / 60
    return "\(hours)h \(minutes)m"
}

/// n<1000 → integer string ("842"); else one decimal + k/M ("10.6k", "1.2M"). Matches
/// `task-display.ts`'s `(n / 1000).toFixed(1)` — both always keep exactly one decimal digit above
/// the 1000 threshold (never strip a trailing ".0").
func formatTokens(_ n: Int) -> String {
    if n < 1000 { return String(n) }
    if n < 1_000_000 { return "\(String(format: "%.1f", Double(n) / 1000))k" }
    return "\(String(format: "%.1f", Double(n) / 1_000_000))M"
}
